tests/wasmfs/wasmfs_opfs.c

```diff
@@ -132,15 +126,59 @@ int main(int argc, char* argv[]) {
   assert(stat_buf.st_size == 42);
   emscripten_console_log("statted while closed again");
 
+  fd = open("/opfs/working/foo.txt", O_RDONLY);
+  assert(fd > 0);
+  emscripten_console_log("opened file in read-only mode");
+
+  char buf2[100] = {};
+  nread = read(fd, buf2, 100);
+  emscripten_console_logf("read message: %s (%d)", buf2, nread);
+  assert(nread == 42);
+  assert(strcmp(buf2, msg) == 0);
+
+  err = fstat(fd, &stat_buf);
+  assert(err == 0);
+  assert(stat_buf.st_size == 42);
+  emscripten_console_log("statted while open in read-only mode");
+
+  int fd2 = open("/opfs/working/foo.txt", O_WRONLY);
+  assert(fd > 0);
+  emscripten_console_log("upgraded open state to read-write");
+
+  lseek(fd2, strlen(msg), SEEK_SET);
+  nwritten = write(fd2, msg, strlen(msg));
+  assert(nwritten == strlen(msg));
+  emscripten_console_logf("wrote message: %s (%d)", msg, nwritten);
+
+  char buf3[100] = {};
+  lseek(fd, 0, SEEK_SET);
+  nread = read(fd, buf3, 100);
+  assert(nread == 42);
+  assert(strcmp(buf3, msg2) == 0);
+  emscripten_console_logf("read message: %s (%d)", buf3, nread);
+
+  close(fd);
+  close(fd2);
+
   fd = open("/opfs/working/foo.txt", O_RDONLY | O_TRUNC);
   assert(fd > 0);
-  emscripten_console_log("truncated while opening");
+  emscripten_console_log("truncated while opening read-only");
+  close(fd);
+
+  err = stat("/opfs/working/foo.txt", &stat_buf);
+  assert(err == 0);
+  assert(stat_buf.st_size == 42);
+  emscripten_console_log("statted after failed truncation");
+
+  fd = open("/opfs/working/foo.txt", O_WRONLY | O_TRUNC);
+  assert(fd > 0);
+  emscripten_console_log("truncated while opening write-only");
   close(fd);
 
   err = stat("/opfs/working/foo.txt", &stat_buf);
   assert(err == 0);
   assert(stat_buf.st_size == 0);
-  emscripten_console_log("statted after truncation");
+  emscripten_console_log("statted after successful truncation");
 
   err = rename("/opfs/working/foo.txt", "/opfs/foo.txt");
   assert(err == 0);
```

Comment:
It might be excessive, but perhaps it's worth a test that goes through all 9 combinations of opening in RDWR/WRONLY/RDONLY and then opens a second time in another mode?

Reply:
Hmm yeah, that is probably warranted...